homeassistant/components/rfxtrx/__init__.py

```diff
@@ -297,9 +302,26 @@ def async_handle_receive(event):
         data_bits = get_device_data_bits(event.device, devices)
         device_id = get_device_id(event.device, data_bits=data_bits)
 
-        # Register new devices
-        if config[CONF_AUTOMATIC_ADD] and device_id not in devices:
-            _add_device(event, device_id)
+        if device_id not in devices:
+            if config[CONF_AUTOMATIC_ADD]:
+                _add_device(event, device_id)
+            else:
+                return
+
+        device_entry = device_registry.async_get_device(
+            identifiers={(DOMAIN, *device_id)},
+            connections=set(),
+        )
+
+        if device_entry is None:
```

Comment:
I don't think we should create a device here. We only raise an event if `fire_event` is `True`. This can only happen if `device_id` is in `devices`. If this is not the case and automatic add is on, the device is added to `devices,` but still then `fire_event` cannot be `True` because for that to happen first options need to be set which reloads the integration. The `async_get_device` should only be called when `fire_event` is `True` and then it should always return a valid device.

Reply:
I think there can be a race condition here, where no entities have been loaded yet. So the device registry may not yet contain this device since it get populated by the platforms.

I think we must add devices to registry in the core logic of the entry. So either here, or earlier in the load of the config entry.

Reply:
That would only happen for devices added through option flow I think? Typically those devices only send and don't receive. Wouldn't it be better in the rare situation this occurs to not fire the event until device exists (ignore if entry is null) and let it be created by the entities?

Reply:
Well it won't be sent it in dispatcher either then. So the entities won't get the data unless we special case that. I just want to avoid these dependencies.

I'm even thinking removing the name/model info from entities and only keep the device link there. Say that it's core/init responsibility to populate the device registry with data.

Reply:
The dispatcher doesn't use device id?

Reply:
I will just populate if device is known for now. I don't really like that the event can sometimes include device id and sometimes not. But I suppose the way events are normally filtered it doesn't matter.